Mathlib/RingTheory/MvPowerSeries/Substitution.lean

```diff
@@ -0,0 +1,435 @@
+  ⟨fun ha ↦
+    {
+      hpow s := (tendsto_pow_of_constantCoeff_nilpotent_iff (a s)).mpr (ha.const_coeff s)
+      tendsto_zero := coeff_zero_iff.mpr ha.coeff_zero },
+    fun ha ↦
+    {
+      const_coeff s := (tendsto_pow_of_constantCoeff_nilpotent_iff (a s)).mp (ha.hpow s)
+      coeff_zero d := (coeff_zero_iff.mp ha.tendsto_zero) d }⟩
+
+theorem HasSubst.hasEval (ha : HasSubst a) :
+    @HasEval σ (MvPowerSeries τ S) _ (@instTopologicalSpace τ S ⊥) a :=
+  hasSubst_iff_hasEval_of_discreteTopology.mp ha
+
+theorem hasSubst_X :
+    HasSubst (fun (s : σ) ↦ (X s : MvPowerSeries σ S)) := by
+  rw [hasSubst_iff_hasEval_of_discreteTopology]
+  exact HasEval.X
+
+theorem hasSubst_zero :
+    HasSubst (fun (_ : σ) ↦ (0 : MvPowerSeries τ S)) := by
+  rw [hasSubst_iff_hasEval_of_discreteTopology]
+  exact HasEval.zero
+
+theorem hasSubst_add {a b : σ → MvPowerSeries τ S}
+    (ha : HasSubst a) (hb : HasSubst b) :
+    HasSubst (a + b) := by
+  rw [hasSubst_iff_hasEval_of_discreteTopology] at ha hb ⊢
+  exact HasEval.add ha hb
+
+@[simp]
+theorem constantCoeff_smul {R : Type*} [Semiring R] {S : Type*} [Semiring S] [Module R S]
+    (φ : MvPowerSeries σ S) (a : R) :
+    constantCoeff σ S (a • φ) = a • constantCoeff σ S φ :=
+  rfl
+
+theorem hasSubst_mul (b : σ → MvPowerSeries τ S)
+    {a : σ → MvPowerSeries τ S} (ha : HasSubst a) : HasSubst (b * a) := by
+  rw [hasSubst_iff_hasEval_of_discreteTopology] at ha ⊢
+  exact HasEval.mul_left b ha
+
+theorem hasSubst_smul (r : MvPowerSeries τ S) {a : σ → MvPowerSeries τ S} (ha : HasSubst a) :
+    HasSubst (r • a) := by convert hasSubst_mul _ ha
+
+/-- Families of mv power series that can be substituted, as an `Ideal` -/
+noncomputable def hasSubst.ideal : Ideal (σ → MvPowerSeries τ S) :=
+  { carrier := setOf HasSubst
+    add_mem' := hasSubst_add
+    zero_mem' := hasSubst_zero
+    smul_mem' := hasSubst_mul }
+
+/-- If σ is finite, then the nilpotent condition is enough for HasSubst -/
+theorem hasSubst_of_constantCoeff_nilpotent [Finite σ]
+    {a : σ → MvPowerSeries τ S} (ha : ∀ s, IsNilpotent (constantCoeff τ S (a s))) :
+    HasSubst a where
+  const_coeff := ha
+  coeff_zero _ := Set.toFinite _
+
+/-- If σ is finite, then having zero constant coefficient is enough for HasSubst -/
+theorem hasSubst_of_constantCoeff_zero [Finite σ]
+    {a : σ → MvPowerSeries τ S} (ha : ∀ s, constantCoeff τ S (a s) = 0) :
+    HasSubst a :=
+  hasSubst_of_constantCoeff_nilpotent (fun s ↦ by simp only [ha s, IsNilpotent.zero])
+
+/-- Substitution of power series into a power series -/
+noncomputable def subst (a : σ → MvPowerSeries τ S) (f : MvPowerSeries σ R) :
+    MvPowerSeries τ S :=
+  MvPowerSeries.eval₂ (algebraMap _ _) a f
+
+variable {a : σ → MvPowerSeries τ S}
+
+/-- Substitution of power series into a power series -/
+noncomputable def substAlgHom (ha : HasSubst a) :
+    MvPowerSeries σ R →ₐ[R] MvPowerSeries τ S :=
+  MvPowerSeries.aeval ha.hasEval
+
+@[simp]
+theorem coe_substAlgHom (ha : HasSubst a) :
```

Comment:
I think it makes more sense to have an applied version `coe_substAlgHom_apply`. Something like
```lean
@[simp]
theorem substAlgHom_apply (ha : HasSubst a) (f : MvPowerSeries σ R) :
    substAlgHom ha f = subst a f := by
```
(we can also keep both versions)

Reply:
indeed.